neqo-http3/src/stream_type_reader.rs

```diff
@@ -49,8 +49,17 @@ impl NewStreamType {
             (HTTP3_UNI_STREAM_TYPE_PUSH, StreamType::UniDi, Role::Client)
             | (WEBTRANSPORT_UNI_STREAM, StreamType::UniDi, _)
             | (WEBTRANSPORT_STREAM, StreamType::BiDi, _) => Ok(None),
-            (H3_FRAME_TYPE_HEADERS, StreamType::BiDi, Role::Server) => Ok(Some(Self::Http)),
-            (_, StreamType::BiDi, Role::Server) => Err(Error::HttpFrame),
+            (_, StreamType::BiDi, Role::Server) => {
+                if stream_type == HTTP3_UNI_STREAM_TYPE_CONTROL
+                    || stream_type == QPACK_UNI_STREAM_TYPE_ENCODER
+                    || stream_type == QPACK_UNI_STREAM_TYPE_DECODER
+                    || stream_type == WEBTRANSPORT_UNI_STREAM
```

Comment:
I think that this is still incorrect, because it's a very tricky piece of code to get right.

In all of these cases, we are reading the first varint from the stream.  For a bidirectional stream, that varint is not a stream type, but a *frame* type.  So the test here is whether the frame is permitted.

For that, we probably want an `HFrame::is_known(u64) -> bool` function so that this can look more or less like this:

```suggestion
                // The "stream_type" for a bidirectional stream is a frame type.
                // We accept WEBTRANSPORT_STREAM (above), and HEADERS, 
                // and we have to ignore unknown types, but any other frame type
                // is bad if we know about it.
                if HFrame::is_known(stream_type) && stream_type != HEADER_FRAME
```

Reply:
Thanks for this. I made the change; it also required changing a test.